[JIT] Some small fixes for NamedTuple

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21839 [JIT] NamedTuple serialization
* #21836 [JIT] Refactor TupleType to take a NamedTupleSpec
* #21835 [JIT] Introduce SerializableType
* **#21813 [JIT] Some small fixes for NamedTuple**


I realized two things:

1) There's an unnecessary ErrorReport if a NamedTuple constructor is called with kwargs. Removed it
2) We don't actually support the default values for fields, so we now throw an error if a NamedTuple definition has them.

Differential Revision: [D15860005](https://our.internmc.facebook.com/intern/diff/D15860005)

Commit:
[JIT] Some small fixes for NamedTuple

Commit:
Update on "[JIT] Some small fixes for NamedTuple"

[JIT] Some small fixes for NamedTuple

gh-metadata: pytorch pytorch 21813 gh/jamesr66a/2/head

Commit:
Update on "[JIT] Some small fixes for NamedTuple"

[JIT] Some small fixes for NamedTuple

gh-metadata: pytorch pytorch 21813 gh/jamesr66a/2/head

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -389,12 +389,17 @@ struct Environment {
-          retval = std::make_shared<script::ClassValue>(class_type);
-        } else if (auto tuple_type = type->cast<TupleType>()) {
-          retval = std::make_shared<script::NamedTupleConstructor>(tuple_type);
+      try {
```
This `try` catches way too many things. What is it suppose to be catching? Generally we avoid all use of try/catch because it messes up debugging.

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -111,6 +111,27 @@ struct PythonResolver : public Resolver {
+            error_msg += kv.first;
+          }
+          error_msg += "]";
+          throw std::runtime_error(error_msg);
```
I would recommend trying to communicate this in any way but throwing a generic exception. For instance, it is probably better to just pass SourceRange to resolveType and let this report the full error message.

zdevito reviewed (approved):
3.6 roentgen. not great. not terrible. should be good to go once it doesn't catch std::exception.

Commit:
Update on "[JIT] Some small fixes for NamedTuple"

[JIT] Some small fixes for NamedTuple

gh-metadata: pytorch pytorch 21813 gh/jamesr66a/2/head

Commit:
Update on "[JIT] Some small fixes for NamedTuple"

[JIT] Some small fixes for NamedTuple

gh-metadata: pytorch pytorch 21813 gh/jamesr66a/2/head

Commit:
Update on "[JIT] Some small fixes for NamedTuple"

[JIT] Some small fixes for NamedTuple

gh-metadata: pytorch pytorch 21813 gh/jamesr66a/2/head
